index.js

```diff
@@ -104,51 +77,279 @@ module.exports = {
 
   errors: errors,
 
-  DEFAULT_HOSTED_NODES: HOSTED_NODES.slice(),
-
-  DEFAULT_HOSTED_WEBSOCKET: HOSTED_WEBSOCKET,
-
-  nodes: {
-    hosted: HOSTED_NODES.slice(),
-    local: null
-  },
-
-  requests: 1,
+  requests: null,
 
   // Hook for transaction callbacks
   txRelay: null,
 
   // Do not call txRelay for these methods
-  excludedFromTxRelay: {},
+  excludedFromTxRelay: null,
 
-  txs: {},
+  txs: null,
 
-  rawTxs: {},
+  rawTxs: null,
 
-  rawTxMaxNonce: -1,
+  rawTxMaxNonce: null,
 
   block: null,
 
   networkID: null,
 
-  notifications: {},
+  notifications: null,
+
+  gasPrice: null,
+
+  configuration: null,
+
+  internalState: null,
+
+  /**
+   * Initiates a connection to Ethereum.  This must be called before any other methods are called.
+   *
+   * @typedef configuration
+   * @type {object}
+   * @property {?string[]} httpAddresses
+   * @property {?string[]} wsAddresses
+   * @property {?string[]} ipcAddresses
+   * @property {?number} connectionTimeout
+   * @property {?function(Error):void} errorHandler - called when an otherwise unhandled asynchronous error occurs during the course of operation.
+   *
+   * @param {!configuration} configuration
+   * @param {!function(?Error):void} initialConnectCallback - if the error parameter is null then the connection was successful
+   * @returns {void}
+   */
+  connect: function (configuration, initialConnectCallback) {
+    this.resetState();
+
+    // overwrite configuration values with user config, throw away unused user config
+    for (var key in this.configuration) {
+      if (configuration[key] === undefined || configuration[key] === null) continue;
```

Comment:
It is unnecessary in this case.  `for .. in ..` only iterates over enumerable properties, which means it excludes things like `toString` on the `Object` prototype.  Also, we are iterating over `this.configuration` which we set/control and it is a simple object with no prototype (beyond the default Object prototype).  Since the Object prototype doesn't have any enumerable properties on it, we can be sure that the we will be iterating only over own properties.